### ISSUE TITLE
CLI / local dev: "supabase db dump" creates broken SQL

### ISSUE THREAD

**Issue by peachp** [supabase/cli#1183]
# Bug report

- [x] I confirm this is a bug with Supabase, not with my own application. 
- [x] I confirm I have searched the [Docs](https://docs.supabase.com), GitHub [Discussions](https://github.com/supabase/supabase/discussions), and [Discord](https://discord.supabase.com). 

## Describe the bug
I have inserted ~130k records into a hosted SB DB using the official JS client without any issues. When I then dump it into a seed.sql on local dev environment and try to reset the DB, it runs into an error.

## To Reproduce
1. Open a local development environment
2. `supabase db dump -f supabase/seed.sql --data-only`:
```
Dumping data from remote database...
pg_dump: warning: there are circular foreign-key constraints among these tables:
Dumped schema to supabase/seed.sql.
```
4. `supabase db reset`:
```
Resetting local database...
Initializing schema...
Applying migration 20230530183107_base.sql...
Applying migration 20230530184344_base.sql...
Applying migration 20230530185011_foreign_keys.sql...
Seeding data supabase\seed.sql...
Restarting containers...
Error: ERROR: syntax error at or near "Villa" (SQLSTATE 42601)
```
5. Inspecting the related record in the DB shows that there is mulit-line TEXT value with some lines starting with `--`

EDIT: When I try `supabase db reset --debug`, it runs for much longer before it stops due to same error.

## Expected behavior
The command `supabase db reset` must restore all records without any errors from `seed.sql` file that was generated by the dump command, esp. considering that they were originally inserted without any issues using the official SB JS client.

## System information
- OS: Windows 11
- Version of supabase-js: n/a
- Version of Node.js: v16.17.0
- Version of Supabase CLI: 1.64.8

## Additional context
A workaround to clean the current records of any "problematic" characters will be too much work. Also I don't want to manipulate future inputs from my users, as we are talking about e.g. generic "Notes" fields here and they should be able to write whatever they like there.

Side note: `supabase db reset` is also very slow. It takes 36s to get to the line ~3000 which causes the error. So if I extrapolate it to ~130,000 lines, it will take around 25min. We haven't gotten into a normal development cycle yet, so depending on how often we'll need to run it, it can be a real issue.

**Comment by sweatybridge** [comment]
Hello @peachp , are you able to isolate the syntax error and share the snippet with us? That would really help with troubleshooting. 

May I also check if you are limiting the memory for docker containers? Seeding is usually very fast because it uses the Postgres binary protocol. The first thing that comes to mind is perhaps your Postgres container is paging data to disk temporarily. 

Alternatively, you may also file a support ticket through the hosted dashboard. I will be able to take a closer look at your project through support.

**Comment by peachp** [comment]
Here you go.

Console:
```
PS C:\Users\foo\Dev\persica> supabase db dump -f supabase/seed.sql --data-only
Dumping data from remote database...
Dumped schema to supabase/seed.sql.
PS C:\Users\foo\Dev\persica> supabase db reset
Resetting local database...
Initializing schema...
Applying migration 20230605135833_schema.sql...
Applying migration 20230606173832_test_table.sql...
Seeding data supabase\seed.sql...
Restarting containers...
Error: ERROR: syntax error at or near "Villa" (SQLSTATE 42601)
Try rerunning the command with --debug to troubleshoot the error.
```

The related record with values as expected, with `-->` at the beginning of some lines, which my users do to organise their notes / pricing info etc.:
![image](https://github.com/supabase/cli/assets/47209189/c2fa8ea4-9187-4f8d-ac61-b522acab259c)


The related SQL piece from seed.sql where you can see the lines with `-->` were removed, causing the next `INSERT INTO` to become a value:
![image](https://github.com/supabase/cli/assets/47209189/c8312204-4fcb-4c14-aaf6-6a8d8cd09e41)

This is the first error when `supabase db reset` aborts. I didn't fix it to see if there are other issues like this further in the file.

I'm using the local Supabase dev / CLI as installed per guide, without adjusting anything special in Docker. I have 32GB of DDR5 with fastest SSD and CPU money could buy a year ago.

I don't know and don't really want to know how to use Docker or "clean" SQL records that are stored in PostgreSQL perfectly fine. I just want to build my app :)

If it helps, I've been trying all sorts of ways to restore a staging / prod DB in local dev, including `pg_dump` and `psql` and Snaplet, but had similar or other issues, so yes, I know our data is not the cleanest, but then again, inserting all these records and then loading into my app using the SB JS client works just fine.

Let me know if you want me to open a support ticket for you to look at my project.

**Comment by sweatybridge** [comment]
Thanks for the detailed explanation. I believe this is a bug with our code to ignore comments. Could you try the following instead?

```bash
supabase db dump -f supabase/seed.sql --data-only --keep-comments
```

**Comment by sweatybridge** [comment]
We found the problem to be with single column inserts flag that we passed to pg_dump. I've release a new version that does batch inserts instead with significantly improved performance. Let me know if it works for you.

```bash
npx supabase@beta db dump -f supabase/seed.sql --data-only
```

**Comment by peachp** [comment]
I will give it a go and let you know. Do I still need the `--keep-comments` flag since we have data with lines starting with `--` ?

**Comment by sweatybridge** [comment]
`--keep-comments` flag is no longer required as it is always true for data-only dumps now.

**Comment by peachp** [comment]
It works now and is fast!

Out of scope of this issue, but related to the whole topic of local dev:

However, after evaluating local dev setup, we decided to use the hosted Supabase after all. Partially because of some issues & limitations with the current Supabase local dev tooling, partially because of our project-specific quirks.

For example we are in pre-go-live stage migrating data from legacy system (Airtable), so we "seed" all the DBs (incl. dev / staging) in same way as we will go live: uploading from a huge JSON file using Supabase JS client.

Once we go live this year, I plan to revisit using local dev. One issue comes to mind though that would still prevent us from using at least the `db dump` and `db reset`:  I want to anonymize / obfuscate the data from prod before I give it to developers. Currently I do it by using `copycat` from Snaplet guys in my script that dumps Airtable records into two versions of the huge JSON 
 file mentioned above: one with real data and one with obfuscated.

It would be great if Supabase CLI had an option to anonymize data. So I (as lead developer) could dump anonymized data from prod into an intermediate hosted Supabase DB, then allow other devs to dump it from there.

**Comment by rusakovic** [comment]
Hello @sweatybridge Do we have a similar error again?

```
supabase -v
1.127.4
supabase db reset --debug
```

Error at
```
2024/01/02 20:50:43 PG Send: {"Type":"Parse","Name":"lrupsc_1_18","Query":"\\n\n--\n-- PostgreSQL database dump\n--\n\n-- Dumped from database version 15.1 (Ubuntu 15.1-1.pgdg20.04+1)\n-- Dumped by pg_dump version 15.5 (Ubuntu 15.5-1.pgdg20.04+1)\n\nSET statement_timeout = 0","ParameterOIDs":null}
2024/01/02 20:50:43 PG Send: {"Type":"Describe","ObjectType":"S","Name":"lrupsc_1_18"}
2024/01/02 20:50:43 PG Send: {"Type":"Sync"}
2024/01/02 20:50:43 PG Recv: {"Type":"ErrorResponse","Severity":"ERROR","SeverityUnlocalized":"ERROR","Code":"42601","Message":"syntax error at or near \"\\\"","Detail":"","Hint":"","Position":1,"InternalPosition":0,"InternalQuery":"","Where":"","SchemaName":"","TableName":"","ColumnName":"","DataTypeName":"","ConstraintName":"","File":"scan.l","Line":1188,"Routine":"scanner_yyerror","UnknownFields":null}
2024/01/02 20:50:43 PG Recv: {"Type":"ReadyForQuery","TxStatus":"I"}
failed to send batch: ERROR: syntax error at or near "\" (SQLSTATE 42601)
```

**Comment by sweatybridge** [comment]
It's a different error, which has been fixed in beta release. Try `npx supabase@beta db dump --data-only`

The next stable release will go out next Tuesday. Since there's a bunch of other changes in beta, we won't be back porting to stable until Tuesday.

**Comment by dhurbhat** [comment]
Hello - I too am getting same error as rusakovic.  (been banging head on this all day...) even after installing `supabase-beta`.

```
2024/01/03 17:02:30 PG Send: {"Type":"Parse","Name":"lrupsc_1_7","Query":"\\n\n--\n-- PostgreSQL database dump\n--\n\n-- Dumped from database version 15.1 (Ubuntu 15.1-1.pgdg20.04+1)\n-- Dumped by pg_dump version 15.5 (Ubuntu 15.5-1.pgdg20.04+1)\n\nSET statement_timeout = 0","ParameterOIDs":null}
2024/01/03 17:02:30 PG Send: {"Type":"Describe","ObjectType":"S","Name":"lrupsc_1_7"}
2024/01/03 17:02:30 PG Send: {"Type":"Sync"}
2024/01/03 17:02:30 PG Recv: {"Type":"ErrorResponse","Severity":"ERROR","SeverityUnlocalized":"ERROR","Code":"42601","Message":"syntax error at or near \"\\\"","Detail":"","Hint":"","Position":1,"InternalPosition":0,"InternalQuery":"","Where":"","SchemaName":"","TableName":"","ColumnName":"","DataTypeName":"","ConstraintName":"","File":"scan.l","Line":1188,"Routine":"scanner_yyerror","UnknownFields":null}
2024/01/03 17:02:30 PG Recv: {"Type":"ReadyForQuery","TxStatus":"I"}
failed to send batch: ERROR: syntax error at or near "\" (SQLSTATE 42601)
```

I'm on MacOS and on reading this have upgraded to beta: 
```
supabase -v
1.131.2
```

Please advise on how to proceed. @sweatybridge   Thank you!

**Comment by sweatybridge** [comment]
@dhurbhat Manually deleting the `\n` characters on the first line also works.

**Comment by alexeigs** [comment]
Same here.

1. Dumping data via `npx supabase@beta db dump -f supabase/seed.sql --data-only --use-copy`
2. `supabase start` leads to:

```
Seeding data supabase/seed.sql...
failed to send batch: ERROR: syntax error at or near "\" (SQLSTATE 42601)
```

**Comment by vasanth-asokan** [comment]
Hello,

This still seems like an issue. I was able to get around it using  `npx supabase@beta db dump --data-only` but that causes a different problem with connecting to remote supabase instances that are now on supavisor. I had to `brew upgrade supabase` to get back to a CLI that would work fine with remote supabase instances. 

Has this fix been merged back into stable non-beta Supabase CLI?

**Comment by rohanrhu** [comment]
I'm backing up my local Supabase for upgrading Supabase CLI and CLI v2.9.6 is still saying this:

```
➜ supabase db dump --local --data-only > supabase/seed.sql

Dumping data from local database...
pg_dump: warning: there are circular foreign-key constraints on this table:
pg_dump: detail: key
pg_dump: hint: You might not be able to restore the dump without using --disable-triggers or temporarily dropping the constraints.
pg_dump: hint: Consider using a full dump instead of a --data-only dump to avoid this problem.
A new version of Supabase CLI is available: v2.12.0 (currently installed v2.9.6)
We recommend updating regularly for new features and bug fixes: https://supabase.com/docs/guides/cli/getting-started#updating-the-supabase-cli
```

Please reopen this issue @sweatybridge

**Comment by sweatybridge** [comment]
@rohanrhu there's no error in your comment, only notice messages from pg_dump.

If you are running into an error, please open a new issue following the template.

**Comment by rohanrhu** [comment]
> [@rohanrhu](https://github.com/rohanrhu) there's no error in your comment, only notice messages from pg_dump.
> 
> If you are running into an error, please open a new issue following the template.

Oh I was searching about the problem that pgsodium has circular foreign keys causing the output `pg_dump: warning: there are circular foreign-key constraints among these tables:` and I've been here because OP's post also has this warning in his output. I've just noticed this issue is about another problem.

But, right now I have just updated my Supabase CLI and ran my local environment but I'm getting weird errors on Studio app. I'm gonna open an issue now.
